src/gui/PianoView.cpp

```diff
@@ -83,39 +88,49 @@ const int LABEL_TEXT_SIZE = 7;      /*!< The height of the key label text */
  *  \param _parent the parent instrument plugin window
  *  \todo are the descriptions of the m_startkey and m_lastkey properties correct?
  */
-PianoView::PianoView( QWidget * _parent ) :
-	QWidget( _parent ),             /*!< Our parent */
-	ModelView( NULL, this ),        /*!< Our view Model */
-	m_piano( NULL ),                /*!< Our piano Model */
-	m_startKey( Key_C + Octave_3*KeysPerOctave ), /*!< The first key displayed? */
-	m_lastKey( -1 )                 /*!< The last key displayed? */
+PianoView::PianoView(QWidget *parent) :
+	QWidget(parent),                 /*!< Our parent */
+	ModelView(nullptr, this),        /*!< Our view Model */
+	m_piano(nullptr),                /*!< Our piano Model */
+	m_startKey(Key_C + Octave_3*KeysPerOctave), /*!< The first key displayed? */
+	m_lastKey(-1),                   /*!< The last key displayed? */
+	m_movedNoteModel(nullptr)        /*!< Key marker which is being moved */
 {
-	if( s_whiteKeyPm == NULL )
+	if (s_whiteKeyPm == nullptr)
 	{
-		s_whiteKeyPm = new QPixmap( embed::getIconPixmap( "white_key" ) );
+		s_whiteKeyPm = new QPixmap(embed::getIconPixmap("white_key"));
 	}
-	if( s_blackKeyPm == NULL )
+	if (s_blackKeyPm == nullptr)
 	{
-		s_blackKeyPm = new QPixmap( embed::getIconPixmap( "black_key" ) );
+		s_blackKeyPm = new QPixmap(embed::getIconPixmap("black_key"));
 	}
-	if( s_whiteKeyPressedPm == NULL )
+	if (s_whiteKeyPressedPm == nullptr)
 	{
-		s_whiteKeyPressedPm = new QPixmap( embed::getIconPixmap( "white_key_pressed" ) );
+		s_whiteKeyPressedPm = new QPixmap(embed::getIconPixmap("white_key_pressed"));
 	}
-	if ( s_blackKeyPressedPm == NULL )
+	if (s_blackKeyPressedPm == nullptr)
 	{
-		s_blackKeyPressedPm = new QPixmap( embed::getIconPixmap( "black_key_pressed" ) );
+		s_blackKeyPressedPm = new QPixmap(embed::getIconPixmap("black_key_pressed"));
+	}
+	if (s_whiteKeyDisabledPm == nullptr)
+	{
+		s_whiteKeyDisabledPm = new QPixmap(embed::getIconPixmap("white_key_disabled"));
+	}
+	if (s_blackKeyDisabledPm == nullptr)
+	{
+		s_blackKeyDisabledPm = new QPixmap(embed::getIconPixmap("black_key_disabled"));
 	}
 
-	setAttribute( Qt::WA_OpaquePaintEvent, true );
-	setFocusPolicy( Qt::StrongFocus );
-	setMaximumWidth( WhiteKeysPerOctave * NumOctaves * PW_WHITE_KEY_WIDTH );
+	setAttribute(Qt::WA_OpaquePaintEvent, true);
+	setFocusPolicy(Qt::StrongFocus);
+	setMaximumWidth(Piano::NumWhiteKeys * PW_WHITE_KEY_WIDTH +
```

Comment:
Maybe leave a comment explaining the reason for this? I can't think of a concise way to do so though.

Black keys are drawn halfway between the preceding and succeeding white note, and contained within their width. Therefore if the first and last keys are white, the maximum required width is the width of all the white notes. However, if the last key is black, this width would cut off the half of that key that would overlap the next white note.

Also, shouldn't this apply if the first note is black as well? Why is the first note guaranteed to be white but not the last one?

Reply:
> Also, shouldn't this apply if the first note is black as well?

I don't think it is currently possible to have the first note black (unless someone changes the black / white definitions in `src/core/Piano.cpp`), since there is no option to make the first octave incomplete, i.e. it always starts from C.

The `PianoView` code is not ready for that possibility either: it always starts drawing from a white key. So if the first key was to be black, it would have to be cut in half, since there is no preceding white key to set as `m_firstKey`.

Anyway, I added a comment, hopefully it makes it a bit clearer.